src/unitxt/metrics.py

```diff
@@ -327,6 +327,11 @@ def score_based_confidence_interval(
             # otherwise, the aggregation_func needs to be applied AFTER resampling the instances;
             #   that is, re-form the groups, calculate the function, and take the mean of the group scores
             aggregation_func = self.average_item_scores
+
+        method = "BCa"
+        if len(instances) > 100:
```

Comment:
What the run time difference and accuracy between BCa and percentile?

Reply:
I do not really know.. I took the advice of @arielge  and @assaftibm , here:
https://github.com/IBM/unitxt/issues/1008#issuecomment-2220794399
and
https://github.com/IBM/unitxt/issues/1008#issuecomment-2222678876

Reply:
I found information about percentile and BCa [here](https://blogs.sas.com/content/iml/2017/07/12/bootstrap-bca-interval.html).
From my understanding (@assaftibm , @arielge , please correct my mistakes..),  
'percentile' works by the following intuitive notion:  make n_resamples of the population [the stream fed to metric],  each such resample is of the size of the population and is built by repeating the following process: independently, and identically distributed, select an item [instance] from the population, with replacement [return of a selected item back to the population].  
Then, given one resample, compute the statistics (the metric) on it, the same way the metric is computed on the original population. 
Then, given n_resamples scores (one per re_sample), return the 0.05 and 0.95 percentile ("quintile" of 100 rather than 5) as the ci borders.

BCa continues the above result of n_resamples scores, by correcting skewness and biases thereof, observed when comparing to the original population.

Reply:
I'm not sure it matters, but I saw somewhere the BCa is the one that is commonly used.   If the difference in runtime is small (I would think the main cost is recalculating the metric and not the BCa vs percentile), then I would keep it simple and not change behavior on some threshold.

So - can you try running twice (with BCA and with percentile) on say 200 instances, and see the difference in runtime?

Reply:
Hi @dafnapension!

Yes - I think your understanding of Percentile bootstrap is correct. 

I would suggest the following:
1. When it's an InstanceMetric, use BCa bootstrap.
2. When it's a GlobalMetric, use Basic bootstrap (or Percentile, see discussion below). 

In 2, I wouldn't condition the type of bootstrap on the size of the dataset and switch between BCa and other lighter bootstrap because it can confuse people that test a full dataset vs. smaller samples of it - they will suddenly get different CIs due to a switch in the bootstrap type. I think it's better to stick to one for consistency.

Regarding Basic vs. Percentile bootstrap.
There is a description [here](https://en.wikipedia.org/wiki/Bootstrapping_(statistics)#Methods_for_bootstrap_confidence_intervals).  I have to admit that I don't fully understand the computation of the Basic bootstrap. However, note that [Scipy documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.bootstrap.html) includes this comment:
> While the 'percentile' method is the most intuitive, it is rarely used in practice. Two more common methods are available, 'basic' (‘reverse percentile’) and 'BCa' (‘bias-corrected and accelerated’); they differ in how step 3 is performed.

So this is the main reason why I would lean towards Basic over Percentile.  It would be easier if we could stick to BCa also in GlobalMetrics but unfortunately that doesn't seem practical.

As a side note, I think it makes sense to ask about the slowdown with BCa on StackOverflow - showing a short self-contained sample code that reproduces  the problem using Scipy's implementation and asking if this is as expected. It's possible that the implementation in Scipy is not optimal.

Reply:
@assaftibm - Are we sure the slow down is due to BcA vs basic (or percentile)?  I think the main slowdown is that if the metric is slow to compute, then all method will call the metric n_samples times.  If n_samples is 100, and the metric takes 1 second, it will take 100 second.

Reply:
About the last note, there's an interesting discussion [here](https://stackoverflow.com/questions/74437089/time-consumption-of-scipys-bootstrap-as-a-function-of-the-number-of-resamples). In our short investigation, @arielge and I also noticed that it is the `jackknife` computation that dominates in BCa. 

In Scipy code, we can see [here](https://github.com/scipy/scipy/blob/87c46641a8b3b5b47b81de44c07b840468f7ebe7/scipy/stats/_resampling.py#L121) an outer for-loop over the dataset, and then [here](https://github.com/scipy/scipy/blob/87c46641a8b3b5b47b81de44c07b840468f7ebe7/scipy/stats/_resampling.py#L129) an inner for-loop over jackknife samples.  The statistic is computed therein.

Reply:
@yoavkatz To clarify, the meaning is that because of the CI computation method (and specifically the `jackknife`), the metric is being called additional times. So it is still a function of the metric being slow, just that with BCa the metric is called many more than `n_resample` times.

Reply:
> I'm not sure it matters, but I saw somewhere the BCa is the one that is commonly used. If the difference in runtime is small (I would think the main cost is recalculating the metric and not the BCa vs percentile), then I would keep it simple and not change behavior on some threshold.
> 
> So - can you try running twice (with BCA and with percentile) on say 200 instances, and see the difference in runtime?

Yes, @yoavkatz , you saw this comment at the documentation of scipy's bootstrap that we use.
Yes, I can try. Running on my own computer, so there is no question of loads by other processes.
I will use the `python -m cProfile -s cumtime xx.py` that I learned from you.

Reply:
@yoavkatz  - to add more info:
* You are right that with global metrics, we need to compute the metric for each resample of the data. So it is `O(n_resamples)` and subject to noticeable slow down when the metric is slow to compute. But this is true no matter what the bootstrap is. So global metrics are slower than instance metrics by definition  as long as CIs are concerned. 
* The new thing (to me) that @arielge saw is that BCa adds a significant slow down by itself.  This is because it attempts to correct for biases in the data, and  part of it is the `jackknife` computation, which computes the metric for each entry in the dataset. It adds at least `O(len(dataset))` calls to the metric (I don't understand from the code whether the number is even higher, it could also be `O(len(dataset)^2)`, see the links I added in the reply to @dafnapension). So even if you reduce the value of `n_resamples` , if your dataset has 10k examples, it will compute the metric 10k times. (or even more) 

So I think that a practical solution would be:
* Instance metrics, use `n_resamples=1k` and `BCa` (or maybe keep it at the default of 10k if the diff is small)
* Global metrics, use `n_resmples=100` and `Basic`
*

Reply:
I wrote a small recipe (I think I am not sufficiently familiar with the new structure thereof, and did several steps "manually":

```python
recipe = StandardRecipe(card="cards.numeric_nlg[task=tasks.generation[metrics=[metrics.meteor[n_resamples=100]]]]",num_demos=0,template_card_index=0,
                        loader_limit=300)
ms = recipe()
ms = {"train" : ms["train"]}
copyapp = Copy(field="target", to_field="prediction")
ms = copyapp(ms)
applypp = ApplyOperatorsField(operators_field="postprocessors")
ms = applypp(ms)
applymetric = ApplyMetric(metric_field="metrics", calc_confidence_intervals=True)
ms = applymetric(ms)
ms_dict = {key: list(ms[key]) for key in ms}
print({key: len(ms_dict[key]) for key in ms_dict})
print(ms_dict["train"][0]["score"])
```
for sanity check: the printed lines are:
```
{'train': 300}
{'global': {'meteor': 0.8932977564065374, 'score': 0.8932977564065374, 'score_name': 'meteor', 'meteor_ci_low': 0.8633536857531038, 'meteor_ci_high': 0.9123895754319261, 'score_ci_low': 0.8633536857531038, 'score_ci_high': 0.9123895754319261}, 'instance': {'meteor': 0.9998177842565598, 'score': 0.9998177842565598, 'score_name': 'meteor'}}
```

I then ran the above script through `cProfile`, in this new branch, having just turned off both conditional additions:
```
        method = "BCa"
        # if len(instances) > 100:
        #     method = "percentile"
```
so that first, only BCa is used, and got the following results:

```
ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      301    0.000    0.000   17.068    0.057 metrics.py:805(process)
        1    0.007    0.007   16.953   16.953 metrics.py:887(compute_instance_scores)
      300    0.005    0.000    3.760    0.013 metrics.py:1326(compute)
        2    0.000    0.000    3.027    1.514 metrics.py:1320(prepare)
        1    0.000    0.000    1.805    1.805 metrics.py:1(<module>)
        1    0.000    0.000    0.114    0.114 metrics.py:295(score_based_confidence_interval)
        3    0.000    0.000    0.100    0.033 metrics.py:342(statistic)
      402    0.001    0.000    0.096    0.000 metrics.py:260(average_item_scores)
      401    0.000    0.000    0.096    0.000 metrics.py:346(<lambda>)
      402    0.003    0.000    0.053    0.000 metrics.py:62(nan_mean)
      402    0.041    0.000    0.041    0.000 metrics.py:269(<listcomp>)
      300    0.002    0.000    0.008    0.000 metrics.py:159(_validate_reference)
      300    0.001    0.000    0.004    0.000 metrics.py:153(_validate_prediction)
      300    0.001    0.000    0.002    0.000 metrics.py:121(_add_score_prefixes_to_score_dict)
      601    0.001    0.000    0.001    0.000 metrics.py:174(get_prediction_type)
      900    0.001    0.000    0.001    0.000 metrics.py:114(_add_score_prefix)
       40    0.000    0.000    0.000    0.000 metrics.py:692(<lambda>)
       34    0.000    0.000    0.000    0.000 metrics.py:478(<lambda>)
        1    0.000    0.000    0.000    0.000 metrics.py:2711(Perplexity)
      300    0.000    0.000    0.000    0.000 metrics.py:1339(<dictcomp>)
        1    0.000    0.000    0.000    0.000 metrics.py:2798(Template)
        1    0.000    0.000    0.000    0.000 metrics.py:284(_all_instance_scores_equal)
        1    0.000    0.000    0.000    0.000 metrics.py:243(new_random_generator)
        3    0.000    0.000    0.000    0.000 metrics.py:371(resample_from_non_nan)
        6    0.000    0.000    0.000    0.000 metrics.py:1186(<lambda>)
        1    0.000    0.000    0.000    0.000 metrics.py:286(<listcomp>)
        1    0.000    0.000    0.000    0.000 metrics.py:1122(MetricPipeline)
        9    0.000    0.000    0.000    0.000 metrics.py:584(<lambda>)
```

Reply:
Repeating all the above, now with the old meteor implementation (as a global HF metric):
First, with BCa  (what we now have in main branch):

```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      301    0.006    0.000   58.748    0.195 metrics.py:473(process)
      702    0.004    0.000   48.132    0.069 metrics.py:540(_compute)
      702    0.032    0.000   48.128    0.069 metrics.py:1195(compute)
        1    0.000    0.000   45.006   45.006 metrics.py:397(compute_global_confidence_intervals)
        3    0.000    0.000   45.003   15.001 metrics.py:403(statistic)
      401    0.011    0.000   44.991    0.112 metrics.py:421(<lambda>)
      401    0.002    0.000   44.929    0.112 metrics.py:406(metric)
        2    0.000    0.000    3.350    1.675 metrics.py:1189(prepare)
        1    0.000    0.000    1.790    1.790 metrics.py:1(<module>)
      401    0.019    0.000    0.019    0.000 metrics.py:422(<listcomp>)
      401    0.017    0.000    0.017    0.000 metrics.py:424(<listcomp>)
      401    0.015    0.000    0.015    0.000 metrics.py:423(<listcomp>)
        1    0.000    0.000    0.003    0.003 metrics.py:130(_validate_references_and_prediction)
      301    0.002    0.000    0.003    0.000 metrics.py:121(_add_score_prefixes_to_score_dict)
      300    0.000    0.000    0.002    0.000 metrics.py:159(_validate_reference)
      903    0.001    0.000    0.001    0.000 metrics.py:114(_add_score_prefix)
      601    0.000    0.000    0.001    0.000 metrics.py:174(get_prediction_type)
       35    0.000    0.000    0.000    0.000 metrics.py:467(<lambda>)
       37    0.000    0.000    0.000    0.000 metrics.py:681(<lambda>)
      300    0.000    0.000    0.000    0.000 metrics.py:153(_validate_prediction)
        1    0.000    0.000    0.000    0.000 metrics.py:2648(Perplexity)
        2    0.000    0.000    0.000    0.000 metrics.py:1177(verify)
        9    0.000    0.000    0.000    0.000 metrics.py:1175(<lambda>)
        1    0.000    0.000    0.000    0.000 metrics.py:2735(Template)
        1    0.000    0.000    0.000    0.000 metrics.py:243(new_random_generator)
        1    0.000    0.000    0.000    0.000 metrics.py:1111(MetricPipeline)
        9    0.000    0.000    0.000    0.000 metrics.py:573(<lambda>)
        3    0.000    0.000    0.000    0.000 metrics.py:365(resample_from_non_nan)

```

Reply:
So, a (unnecessarily) global - is clearly out of consideration.  I next checked the difference between both instance-implementation, across BCa vs percentile, when we increase n_resamples from 100 (as above) to 1000.

Reply:
And last, increased the number of instances from 300 to 1200.
First, BCa:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1201    0.003    0.000   53.394    0.044 metrics.py:805(process)
        1    0.030    0.030   48.052   48.052 metrics.py:887(compute_instance_scores)
     1200    0.026    0.000   14.935    0.012 metrics.py:1326(compute)
        1    0.003    0.003    5.336    5.336 metrics.py:295(score_based_confidence_interval)
        3    0.000    0.000    5.153    1.718 metrics.py:342(statistic)
        1    0.000    0.000    5.079    5.079 metrics.py:1(<module>)
     2201    0.007    0.000    5.079    0.002 metrics.py:346(<lambda>)
     2202    0.062    0.000    5.075    0.002 metrics.py:260(average_item_scores)
     2202    3.603    0.002    3.603    0.002 metrics.py:269(<listcomp>)
        2    0.000    0.000    3.516    1.758 metrics.py:1320(prepare)
     2202    0.112    0.000    1.410    0.001 metrics.py:62(nan_mean)
     1200    0.008    0.000    0.045    0.000 metrics.py:159(_validate_reference)
     1200    0.006    0.000    0.015    0.000 metrics.py:153(_validate_prediction)
     1200    0.010    0.000    0.013    0.000 metrics.py:121(_add_score_prefixes_to_score_dict)
        1    0.000    0.000    0.006    0.006 metrics.py:284(_all_instance_scores_equal)
        1    0.004    0.004    0.004    0.004 metrics.py:289(<listcomp>)
     2401    0.003    0.000    0.003    0.000 metrics.py:174(get_prediction_type)
        1    0.003    0.003    0.003    0.003 metrics.py:286(<listcomp>)
     3600    0.002    0.000    0.002    0.000 metrics.py:114(_add_score_prefix)
     1200    0.001    0.000    0.001    0.000 metrics.py:1339(<dictcomp>)
       34    0.000    0.000    0.001    0.000 metrics.py:478(<lambda>)
       40    0.000    0.000    0.001    0.000 metrics.py:692(<lambda>)
        1    0.000    0.000    0.001    0.001 metrics.py:2711(Perplexity)
        1    0.000    0.000    0.001    0.001 metrics.py:2798(Template)
        1    0.000    0.000    0.000    0.000 metrics.py:243(new_random_generator)
        6    0.000    0.000    0.000    0.000 metrics.py:1186(<lambda>)
        3    0.000    0.000    0.000    0.000 metrics.py:371(resample_from_non_nan)
        9    0.000    0.000    0.000    0.000 metrics.py:584(<lambda>)
```
and then percentile:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1201    0.002    0.000   60.528    0.050 metrics.py:805(process)
        1    0.041    0.041   58.954   58.954 metrics.py:887(compute_instance_scores)
     1200    0.030    0.000   18.488    0.015 metrics.py:1326(compute)
        2    0.000    0.000    3.005    1.502 metrics.py:1320(prepare)
        1    0.000    0.000    1.900    1.900 metrics.py:1(<module>)
        1    0.004    0.004    1.570    1.570 metrics.py:295(score_based_confidence_interval)
        1    0.000    0.000    1.490    1.490 metrics.py:342(statistic)
     1000    0.002    0.000    1.469    0.001 metrics.py:346(<lambda>)
     1001    0.016    0.000    1.469    0.001 metrics.py:260(average_item_scores)
     1001    1.101    0.001    1.101    0.001 metrics.py:269(<listcomp>)
     1001    0.022    0.000    0.352    0.000 metrics.py:62(nan_mean)
     1200    0.010    0.000    0.045    0.000 metrics.py:159(_validate_reference)
     1200    0.007    0.000    0.017    0.000 metrics.py:153(_validate_prediction)
     1200    0.008    0.000    0.015    0.000 metrics.py:121(_add_score_prefixes_to_score_dict)
     3600    0.005    0.000    0.005    0.000 metrics.py:114(_add_score_prefix)
     2401    0.002    0.000    0.003    0.000 metrics.py:174(get_prediction_type)
        1    0.000    0.000    0.002    0.002 metrics.py:284(_all_instance_scores_equal)
     1200    0.002    0.000    0.002    0.000 metrics.py:1339(<dictcomp>)
        1    0.001    0.001    0.001    0.001 metrics.py:286(<listcomp>)
        1    0.001    0.001    0.001    0.001 metrics.py:289(<listcomp>)
       40    0.000    0.000    0.000    0.000 metrics.py:692(<lambda>)
       34    0.000    0.000    0.000    0.000 metrics.py:478(<lambda>)
        1    0.000    0.000    0.000    0.000 metrics.py:243(new_random_generator)
        1    0.000    0.000    0.000    0.000 metrics.py:2711(Perplexity)
        6    0.000    0.000    0.000    0.000 metrics.py:1186(<lambda>)
        1    0.000    0.000    0.000    0.000 metrics.py:2798(Template)
        9    0.000    0.000    0.000    0.000 metrics.py:584(<lambda>)
```

Reply:
In all my experiments, because of lack of real data, I used prediction=target (first reference). I think it only change a constant time of computing each instance score, so I allowed myself..
In my humble opinion, in the numbers of instances and resamples that I see in the cards, we can stick to the said to be most commonly used BCa.   Just remember to check again if we grow dramatically.

Reply:
Thanks for the detailed analysis.   BCa seems to increase the time about 10%-20% in instance metrics, but more than doubles the time in global metrics.

Reply:
So do we agree to keep BCa always? I don't think we have many more global metrics that take a long time - and it's the simplest approach.

Reply:
I, too, think so. Leave the change for a future consideration, if and when needed.
@elronbandel also asked that I simply copy over the code from what we find in HF to an in-house instance metric, for a further speedup.

Reply:
I think the problem will appear with more complex global metrics like CorpusBLEU, but I agree it can be dealt with in  a separate PR.